plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,435 @@
+import           Data.Coerce
+import qualified Data.List.NonEmpty      as NE
+import qualified Data.Map                as M
+import           Data.Semigroup.Foldable
+import qualified Data.Set                as S
+import           GHC.Exts
+import qualified PlutusCore              as PLC
+import qualified PlutusCore.Constant     as PLC
+import qualified PlutusCore.Name         as PLC
+import           PlutusIR
+import           PlutusIR.Purity
+import           PlutusIR.Subst
+
+
+{- Note [Let Floating pass]
+
+The goal of this pass is to move (float) let-bindings as outwards as possible,
+without breaking the scoping & meaning of the original PIR term.
+
+This transformation (a.k.a. full laziness), together with a possible implementation
+is described in Peyton Jones, Simon, Will Partain, and Andre Santos. "Let-Floating: Moving Bindings to Give Faster Programs."
+In Proceedings of the First ACM SIGPLAN International Conference on Functional Programming, 1-12.
+ICFP '96. New York, NY, USA: ACM, 1996. https://doi.org/10.1145/232627.232630.
+
+An implementation, as described in the paper, is comprised of two "passes":
+
+1) a "mark" pass to travers the term tree and
+  - in case of lam/Lam, mark this lam/Lam name with current depth, and
+    increase the depth for the lam/Lam's-abstraction body term and recurse.
+  - in case of a Letrecgroup, collect the free term&type variables and mark every let-introduced name
+    with the maximum depth among all the free variables (the free variables should be already marked)
+  - in case of letnonrec group, you can treat it the same as (letrec g in letrec gs)
+
+2) a "float-back" pass which, given the collected marks,
+   traverses the term tree again and whenever a let(rec or nonrec) is encountered,
+   decides locally if it is worth to float the current let outwards at its marked depth.
+   If yes, the let-group's binding is floated exactly outside a lambda abstraction that has lam_depth=let_depth+1
+
+There are some  differences with the paper's described implementation above, namely:
+
+a) we use 3 passes. the 1st pass is similar to the original; a second pass
+"cleans" the term from all the to-be-floated lets and stores them in a separate table.
+the 3rd pass is responsible to float back the removed lets inside the reducted (cleaned) term
+according to their markers. So we use an extra pass because we float back lets in a global fashion,
+instead of deciding locally.
+
+b) Since the 3rd (float-back) pass operates on a reducted (cleaned) term, we have lost
+the original location of the lets, so we cannot float them "right outside" the **maximum-independent lambda-abstraction**,
+but we float them "right inside" the maximum **dependent** lambda-abstraction's body. This has the downside
+of allocating&holding the lets for longer than needed, but will not alter the meaning of the original PIR term.
+
+c) Since PIR has strict (compared to the paper's lazy-only lang), we have to make
```

Comment:
I don't think I understand this statement.

Is the _parent_:
- the parent lambda closure, or
- the parent LET node?

Comment mentions that the parent is a parent LET node, but I don't understand how a parent can be floated if it has a side-effectfull term as it's child?  Do you mean that it can be floated down?

Reply:
Consider this example:

`.... let (nonstrict) x1=  (let (strict) x2 = error in rhs1) in body1`

the parent is x1 and it is floatable
the child is x2 and is unmovable

The x2 will not move with respect to x1, but its *original absolute program location* may change. Does this make sense?

Reply:
Ah, right, forgot about the non-strict let bindings.  Yeah, that makes sense now.